### PR TITLE
WIP: Process .cma's, .cmi's and toplevel js files

### PR DESCRIPTION
**Pull request by gpetiot**
Upstreaming patches of https://github.com/ocaml-doc/voodoo/compare/main...jonludlam:voodoo:jsoo

### CODE REVIEW

File: src/voodoo-prep/main.ml [hunk 1]
```diff
@@ -43,12 +43,39 @@ module Prep = struct
...
+      "Provide extra universes as 'package=universe id' couples. These will \
+       not be prepped but can be referenced by jsoo toplevels."
+    in
+    Arg.(
+      value
+      & opt (list (pair ~sep:':' string string)) []
+      & info [ "v"; "extra-universes" ] ~doc)
+
+  let cmd = Term.(const prep $ lib_dir $ switch $ universes $ extra_universes)
   let info = Term.info "prep" ~doc:"Prep a directory tree for compiling"
```
Comment (sabine) [inline src/voodoo-prep/main.ml:57]: @jonludlam I am sure there is a very good high-level explanation for why we now need `extra_universes` as a parameter to voodoo-prep.

Can you elaborate on that?

**Review by sabine — COMMENTED** [review]
I like it. Just some small observations on this WIP PR at this point:

- dependency to `js_of_ocaml` needs to be added
- maybe we can try to encapsulate the parts of the code that will be reusable? I don't know how much of this is actually reusable for other projects that need to build toplevels (maybe it will ultimately live in odoc? or maybe in a separate project?)

File: test/can-render-org-files.t [hunk 2]
```diff
@@ -2,11 +2,1931 @@ Generate the base documentation
...
+              -o jsoo/universes/32/ctypes/0.20.2/ctypes-top.cma.worker.js
+  Cma: js_of_ocaml prep/universes/32/ctypes/0.20.2/lib/ctypes/ctypes-top.cma
+         --pretty --wrap-with-fun ctypes-top_dot_cma -o
+         jsoo/universes/32/ctypes/0.20.2/ctypes-top.cma.js
+  js_of_ocaml: option '--wrap-with-fun': must be empty or a valid JavaScript
+               identifier
+  Usage: js_of_ocaml compile [OPTION]… [JS_FILES]… PROGRAM
+  Try 'js_of_ocaml compile --help' or 'js_of_ocaml --help' for more information.
+  voodoo-jsoo: internal error, uncaught exception:
+               Failure("Error: run ['js_of_ocaml'\n     'prep/universes/32/ctypes/0.20.2/lib/ctypes/ctypes-top.cma' '--pretty'\n     '--wrap-with-fun' 'ctypes-top_dot_cma' '-o'\n     'jsoo/universes/32/ctypes/0.20.2/ctypes-top.cma.js']: exited with 124")
```
Comment (gpetiot) [inline test/can-render-org-files.t:1921]: @jonludlam do you know what is supposed to be the correct value for the `--wrap-with-fun` arg here? Should it depend on something in https://github.com/jonludlam/js_top_worker?

Comment (jonludlam) [reply test/can-render-org-files.t:1921]: looks like it needs a `sed s/-/_/g` or the like.

**Comment by gpetiot** [comment]
@jonludlam @sabine @trefis Here are more details about the js failures at runtime.

I added some logs to `voodoo-prep` and `voodoo-prep jsoo`:
```
main.exe: [INFO] Processing prep/universes/1/astring/0.8.5/astring.cma.worker.js.jstop
Warning: your program contains effect handlers; you should probably run js_of_ocaml with option '--enable=effects'
main.exe: [INFO] Produced jsoo/universes/1/astring/0.8.5/astring.cma.js
main.exe: [INFO] Wrapping function: astring_dot_cma
main.exe: [INFO] Produced jsoo/universes/1/astring/0.8.5/astring_top.cma.js
main.exe: [INFO] Wrapping function: astring_top_dot_cma
main.exe: [INFO] Processing prep/universes/1/astring/0.8.5/astring_top.cma.worker.js.jstop
Warning: your program contains effect handlers; you should probably run js_of_ocaml with option '--enable=effects'
main.exe: [INFO] toplevel: { package: (1,astring,0.8.5); bc_exe_fpath: lib/astring/astring_top.cma; js_files: [] }
main.exe: [INFO] toplevel: { package: (1,astring,0.8.5); bc_exe_fpath: lib/astring/astring.cma; js_files: [] }
```

These files are then copied into the playground assets (I ignore the astring_top.cma.worker.js.jstop on purpose I'm only interested in loading the cma files dynamically).

The toplevel is initialised with:
```ocaml
Toprpc.init rpc
      {
        Toplevel_api.cmi_urls = cmi_urls ();
        cmas =
          [
            { url = "stdlib/astring.cma.js"; fn = "astring_dot_cma" };
            { url = "stdlib/astring_top.cma.js"; fn = "astring_top_dot_cma" };
          ];
      }
```

And finally the errors are:
Screenshot 2023-08-21 at 18 43 51

**Pull request closed without merging**